spinalcordtoolbox/deepseg/core.py

```diff
@@ -0,0 +1,144 @@
+DEFAULT_FILL_HOLES = True
+
+
+class ParamDeepseg:
+    """
+    Parameters for deepseg module.
+    """
+    def __init__(self):
+        self.threshold = None
+        self.keep_largest_object = None
+        self.fill_holes = None
+        self.output_suffix = '_seg'
+        self.remove_temp_files = 1
+        self.verbose = 1
+        # TODO: add threshold, keep_big_obj
+
+
+class PostProcessing:
+    """
+    Deals with post-processing of the segmentation. Consider param (i.e. user's flag) with more priority than
+    metadata (i.e. from model's json file).
+    """
+    def __init__(self, param, metadata):
+        """
+        :param param: class ParamDeepseg: Defined by user's parameter
+        :param metadata: dict: From model's json metadata
+        """
+        self.param = param
+        self.metadata = metadata
+
+    def threshold(self, nii_seg):
+        """
+        Threshold the prediction. For no prediction, set 'threshold' to 0.
+        """
+        if self.param.threshold:
+            thr = self.param.threshold
+        else:
+            if 'threshold' in self.metadata:
+                thr = self.metadata.threshold
+            else:
+                logger.warning("'threshold' is not defined in the model json file. Using threshold of: {}".format(
+                    DEFAULT_THRESHOLD))
+                thr = DEFAULT_THRESHOLD
+        if thr:
+            nii_seg = imed.postprocessing.threshold_predictions(nii_seg, thr)
+        return nii_seg
+
+    def keep_largest_object(self, nii_seg):
+        """
+        Only keep largest object
+        """
+        # TODO: This if/elif below is ugly. Cannot think of something better for now...
+        do_process = DEFAULT_KEEP_LARGEST_OBJECT
+        if self.param.keep_largest_object is True:
+            do_process = True
+        elif self.param.keep_largest_object is None:
+            if 'keep_largest_object' in self.metadata:
+                do_process = self.metadata.keep_largest_object
+        if do_process:
+            # Make sure input is binary
+            if np.array_equal(nii_seg.get_fdata(), nii_seg.get_fdata().astype(bool)):
+                # Fetch axis corresponding to superior-inferior direction
+                # TODO: move that code in image
+                affine = nii_seg.get_header().get_best_affine()
+                code = nib.orientations.aff2axcodes(affine)
+                if 'I' in code:
+                    axis_infsup = code.index('I')
+                elif 'S' in code:
+                    axis_infsup = code.index('S')
+                else:
+                    raise ValueError(
+                        "Neither I nor S is present in code: {}, for affine matrix: {}".format(code, affine))
+                nii_seg = imed.postprocessing.keep_largest_object_per_slice(nii_seg, axis=axis_infsup)
+            else:
+                logger.warning("Algorithm 'keep largest object' can only be run on binary segmentation. Skipping.")
```

Comment:
Could this warning be an exception? "Errors should never pass silently. Unless explicitly silenced." advocates the zen of python.

Reply:
well, technically it is not an error but just a warning that this step will be skipped. If we stop the program here, i anticipate that all users who try the flag "-thr 0" will see a crash, and since most users to read their crash report, they will wrongly think that "-thr 0" doesn't work...  so i'd rather give them an output and informing the small percentage of users who actually read reports that some algo have not run because of -thr 0.
@charleygros do you have any opinion?

Reply:
I would agree that this is more a warning than an error, and would then agree to skip it.
Another approach would be to call the thresholding with a default value (eg 10e-3) in case the object is not binary (ie thr=0) when entering in `keep_largest_object` and then use the `mask_prediction` function to remove the small objects from the soft pred based on the binary output of `keep_largest`.

Reply:
> Another approach would be to call the thresholding with a default value (eg 10e-3) in case the object is not binary (ie thr=0) when entering in keep_largest_object and then use the mask_prediction function to remove the small objects from the soft pred based on the binary output of keep_largest.

yup i like this idea!

Reply:
This should be implemented in ivadomed though, issue opened: https://github.com/neuropoly/ivado-medical-imaging/issues/205